### PR TITLE
Better document what a spherical shell with topography does with normal vectors.

### PR DESCRIPTION
**Pull request by bangerth**
As discussed this morning. I think this comment adequately describes what we want to happen.

I'll note that I think there is still something wrong, but will address this in a separate PR.

### CODE REVIEW

**Comment by bangerth** [comment]
Part of #5421.

File: include/aspect/geometry_model/spherical_shell.h [hunk 1]
```diff
@@ -95,6 +95,18 @@ namespace aspect
 
           /**
            * @copydoc Manifold::normal_vector()
+           *
+           * We fudge here, but for a good reason. What the function is supposed
+           * to compute is the normal vector to the surface. This *should* be the
+           * normal to the surface with topography, but instead we return the
+           * normal to the undeformed surface -- i.e., the radial direction. This
+           * is, in particular, used to compute no-flux boundary conditions,
```
Comment (tjhei) [inline include/aspect/geometry_model/spherical_shell.h:103]: My understanding is that VectorTools::compute_no_normal_flux_constraints() ignores the Manifold normal while VectorTools::compute_no_normal_flux_constraints_on_level() uses it. I should document this inside deal.II...

File: include/aspect/geometry_model/spherical_shell.h [hunk 2]
```diff
@@ -95,6 +95,18 @@ namespace aspect
...
+           * for which we want to impose something resembles a boundary
+           * condition that allows for plate-like motion -- that is, we need
```
Comment (tjhei) [inline include/aspect/geometry_model/spherical_shell.h:105]: Is this always the right thing to do or just for the problem we are looking at right now? @jdannberg ?

File: source/geometry_model/spherical_shell.cc [hunk 3]
```diff
@@ -210,7 +210,21 @@ namespace aspect
       normal_vector(const typename Triangulation<dim, dim>::face_iterator &face,
                     const Point<dim> &p) const
       {
-        // TODO: Deal with pull back and push forward
+        // We fudge here, but for a good reason. What the function is supposed
```
Comment (tjhei) [inline source/geometry_model/spherical_shell.cc:213]: I prefer not to duplicate comments like this and rather refer to the other location.

File: include/aspect/geometry_model/spherical_shell.h (hunk 1 above)
Comment (bangerth) [reply include/aspect/geometry_model/spherical_shell.h:103]: Yes please!

File: source/geometry_model/spherical_shell.cc (hunk 3 above)
Comment (bangerth) [reply source/geometry_model/spherical_shell.cc]: Fair point. Fixed.

File: include/aspect/geometry_model/spherical_shell.h [hunk 4]
```diff
@@ -95,6 +95,18 @@ namespace aspect
 
           /**
            * @copydoc Manifold::normal_vector()
+           *
+           * We fudge here, but for a good reason. What the function is supposed
+           * to compute is the normal vector to the surface. This *should* be the
+           * normal to the surface with topography, but instead we return the
+           * normal to the undeformed surface -- i.e., the radial direction. This
+           * is, in particular, used to compute no-flux boundary conditions,
+           * for which we want to impose something resembles a boundary
```
Comment (jdannberg) [inline include/aspect/geometry_model/spherical_shell.h:104]:
```suggestion
           * for which we want to impose a boundary
```
Or at least I found the wording as it is right now confusing.

File: include/aspect/geometry_model/spherical_shell.h (hunk 2 above)
Comment (jdannberg) [reply include/aspect/geometry_model/spherical_shell.h:105]: I could see cases where one would not want this. For example, if we model flow through a pipe and (ab)use the initial topography plugin to define the upper boundary of the pipe. So, maybe someone does not want to include the lithosphere in their model and only looks at flow through the asthenosphere, and then the initial topography corresponds to the lithosphere depth. 

But that is not a very common case, I would think.

File: source/geometry_model/spherical_shell.cc [hunk 5]
```diff
@@ -210,7 +210,13 @@ namespace aspect
...
                     const Point<dim> &p) const
       {
-        // TODO: Deal with pull back and push forward
+        // We calculate radial, rather than *normal*, vectors here if a face is
+        // at the boundary. This is as described in the documentation of this
+        // function.
+
+        // TODO: If this turns out to ever be wrong, we can always make it
+        //   a parameter whether we want this or the "geometrically correct"
+        //   behavior.
```
Comment (jdannberg) [inline source/geometry_model/spherical_shell.cc:219]: I think there are cases where one would not want this, see above. So maybe

```suggestion
        // TODO: Add an input parameter that allows it to select
        // whether we want this or the "geometrically correct"
        // behavior.
```

**Comment by bangerth** [comment]
Addressed as requested.

**Review by jdannberg — APPROVED** [review]

**Comment by gassmoeller** [comment]
the Jenkins tester passed all tests but then failed during the final cleanup stage. That looks more like a problem with the Jenkins tester, but I will retrigger it one time to make sure. Afterwards this is good to go.

**Pull request merged**